### PR TITLE
Support for ROS 2

### PR DESCRIPTION
**Pull request by micragz**
This commit updates the zivid driver to support ROS2.

See the CHANGELOG.md for user-facing changes and see the README.md for a general description of the ROS2 driver.

### CODE REVIEW

File: CHANGELOG.md
Comment (micragz) [inline CHANGELOG.md:5]: Reminder to update this before merge.

File: zivid_samples/src/sample_capture_with_settings_from_file.cpp
Comment (apartridge) [inline zivid_samples/src/sample_capture_with_settings_from_file.cpp:26]: Should we replace these terminate's with exceptions?

Comment (micragz) [reply zivid_samples/src/sample_capture_with_settings_from_file.cpp:26]: We can do that, I didn't want to replicate the error messages twice, and felt this was simpler than adding another function to every sample. But I'm open to doing that as well.

Comment (apartridge) [reply zivid_samples/src/sample_capture_with_settings_from_file.cpp:26]: OK, I think we want to use exceptions to follow best-practices and have a clean shutdown, but we don't want to duplicate the text of course. So I think making a small helper function is fine but I would duplicate it so that each sample is standalone.

Comment (micragz) [reply zivid_samples/src/sample_capture_with_settings_from_file.cpp]: Fixed in 3d9da486dbef8d393f20a2ee589771e73dcecd8c

File: zivid_camera/test/test_zivid_camera.cpp
Comment (micragz) [inline zivid_camera/test/test_zivid_camera.cpp:193]: The `[&]` capture here looks suspicious, the `w` object will go out of scope after returning from the function. Perhaps store the reference to the impl itself?

File: zivid_samples/launch/zivid_camera_with_serial_number.launch
Comment (apartridge) [inline zivid_samples/launch/zivid_camera_with_serial_number.launch:2]: set X to SERIAL?

File: README.md
Comment (apartridge) [inline README.md:10]: We must update this. Also point to ros1-master for ROS1 code.

File: zivid_camera/test/test_zivid_camera.cpp
Comment (micragz) [inline zivid_camera/test/test_zivid_camera.cpp:1135]: I think we should extract the temporary file code from `TmpFile` to get a platform-independent temporary path. Applies to all `/tmp` paths.

**Review by micragz — COMMENTED** [review]
A few comments, mostly minor. Nice work!

File: zivid_camera/test/test_zivid_camera.cpp
Comment (micragz) [inline zivid_camera/test/test_zivid_camera.cpp:131]: Use 

> C:\ProgramData\Zivid\FileCameraZivid2M70.zfc

On Windows.

File: zivid_camera/test/test_zivid_camera.cpp
Comment (micragz) [inline zivid_camera/test/test_zivid_camera.cpp:1282]: Use the path from the variable above.

Comment (micragz) [reply zivid_camera/test/test_zivid_camera.cpp]: 168d43bdd1ce3b1c131ff6b49e12365868f821eb

File: zivid_camera/test/test_zivid_camera.cpp
Comment (micragz) [reply zivid_camera/test/test_zivid_camera.cpp]: 168d43bdd1ce3b1c131ff6b49e12365868f821eb

File: zivid_camera/test/test_zivid_camera.cpp
Comment (micragz) [reply zivid_camera/test/test_zivid_camera.cpp]: 168d43bdd1ce3b1c131ff6b49e12365868f821eb

File: zivid_camera/test/test_zivid_camera.cpp
Comment (micragz) [reply zivid_camera/test/test_zivid_camera.cpp]: 138913e9a424e4669c3685b47d7a3ce3bb239c92

File: zivid_samples/launch/zivid_camera_with_serial_number.launch
Comment (micragz) [reply zivid_samples/launch/zivid_camera_with_serial_number.launch:2]: Maybe better to remove the default? It makes little sense to invoke this without the serial number. If I do this, I get the following:

```
> ros2 launch zivid_samples zivid_camera_with_serial_number.launch
[INFO] [launch]: All log files can be found below C:\Users\MichaelRPRagazzon\.ros\log\2024-07-22-16-10-01-579606-DESKTOP-3CHQG3M-18084
[INFO] [launch]: Default logging verbosity is set to INFO
[ERROR] [launch]: Caught exception in launch (see debug for traceback): Included launch description missing required argument 'serial_number' (description: 'no description given'), given: []
```

If we add a description too, e.g. `<arg name="serial_number" description="Serial number of camera to connect to"/>`, we get this output:

```
[ERROR] [launch]: Caught exception in launch (see debug for traceback): Included launch description missing required argument 'serial_number' (description: 'Serial number of camera to connect to'), given: []
```

File: README.md
Comment (micragz) [reply README.md]: b11482c9b92690a4085c2f7c24589c0b05cc1a15

File: README.md
Comment (micragz) [reply README.md]: b11482c9b92690a4085c2f7c24589c0b05cc1a15

File: README.md
Comment (micragz) [reply README.md]: b11482c9b92690a4085c2f7c24589c0b05cc1a15

File: README.md
Comment (micragz) [reply README.md]: b11482c9b92690a4085c2f7c24589c0b05cc1a15

File: zivid_samples/launch/zivid_camera_with_serial_number.launch
Comment (micragz) [reply zivid_samples/launch/zivid_camera_with_serial_number.launch]: f43ee1192b8df8861a4394c90bae795000decff9

File: zivid_camera/src/zivid_camera.cpp
Comment (micragz) [reply zivid_camera/src/zivid_camera.cpp:239]: Yeah, that makes sense to me. I made an effort at implementing this, let me know if it looks alright.

abfb5eba7dd9633f15cc2fde0f7ccdd30dd57f12

**Comment by micragz** [comment]
I'm happy with this from my perspective. I left some comments open that you might want to take a look at @apartridge. I can't approve this since I opened the PR, but consider it approved from my view :)

File: zivid_camera/src/zivid_camera.cpp
Comment (apartridge) [inline zivid_camera/src/zivid_camera.cpp:530]: These variables should always be non-null? So we don't need the if-checks.

File: README.md
Comment (apartridge) [reply README.md:28]: Yes we should also mention U24.

File: README.md
Comment (apartridge) [reply README.md:103]: Not sure about this. I think the same warning is in ROS1 driver, can you check that?

File: zivid_camera/test/test_zivid_camera.cpp
Comment (apartridge) [inline zivid_camera/test/test_zivid_camera.cpp:169]: missing throw

File: zivid_camera/src/zivid_camera.cpp
Comment (apartridge) [inline zivid_camera/src/zivid_camera.cpp:226]: can remove this `using`

File: zivid_camera/src/zivid_camera.cpp
Comment (micragz) [reply zivid_camera/src/zivid_camera.cpp:530]: There is an initialization order issue complicating these matters. The callback is set before the settings controllers are constructed, thus they will be null for the first few callbacks.

We could move the callback to the bottom, then the default parameters being set won't be logged. We could also initialize all of zivid before the ros things, but that might have its own problems, e.g. if the zivid application won't start because of a bad configuration.

File: zivid_camera/test/test_zivid_camera.cpp
Comment (micragz) [reply zivid_camera/test/test_zivid_camera.cpp:169]: 446c49665bbcb609a5c5ed913f49727a03db9142

File: zivid_camera/src/zivid_camera.cpp
Comment (micragz) [reply zivid_camera/src/zivid_camera.cpp:226]: 809f259825b4e23f1e089907fcbf3ef56eaf8cf9

File: zivid_camera/src/zivid_camera.cpp
Comment (micragz) [reply zivid_camera/src/zivid_camera.cpp:199]: 5cca7670772f09886da6001cb9a7aec53009273c

File: README.md
Comment (micragz) [reply README.md]: 773f841a25ebc9226e718f70590dfbe34df80ab8

File: README.md
Comment (apartridge) [inline README.md:69]: must remove the `-b ros2-preview`

File: zivid_camera/src/zivid_camera.cpp
Comment (apartridge) [inline zivid_camera/src/zivid_camera.cpp:68]: could you change this method to camel-case bigEndian() to match the other funcitons?

File: zivid_camera/src/zivid_camera.cpp
Comment (apartridge) [inline zivid_camera/src/zivid_camera.cpp:270]: It may be this line (and `cached_settings_.reset();` above) is not needed? Will the onSetParameter() line below be called when node_.set_parameter() is called? If so, then it may be best to remove these cached_settings lines, and let the regular mechanism deal with the caching.

File: README.md
Comment (micragz) [reply README.md]: df24da9329e235dcb30399b53eb6c1781f73d634

File: zivid_camera/src/zivid_camera.cpp
Comment (micragz) [reply zivid_camera/src/zivid_camera.cpp:68]: df24da9329e235dcb30399b53eb6c1781f73d634

File: zivid_camera/src/zivid_camera.cpp
Comment (micragz) [reply zivid_camera/src/zivid_camera.cpp]: Yeah, that's true. I was thinking the reset might be needed in case of exceptions in `serializeZividDataModel`, but that's not really a case we want to handle, so better let it propagate out.

And the last assignment was mostly because it seemed more efficient than interpreting the string. But this isn't really a case where we need to optimize. So I removed both of these. Also fixed up some debug logging.

56ce1a0a6e312f046e94aa23ae98be8c949ad6ac

**Review by apartridge — COMMENTED** [review]
LGTM, if you can squash the fixups then I will approve

**Review by apartridge — APPROVED** [review]

**Pull request merged**
